### PR TITLE
fix(python): List[Datatype] always return bool

### PR DESCRIPTION
**Pull request by Mehrbod2002**
It's related to #6206 ,
There was some conflict between files.

### CODE REVIEW

**Comment by ghuls** [comment]
I think it needs to be fixed differently.

If I patch dtype list representation, printing the dataframe with a list column reflects this modified dtype.
So for `pl.List(pl.Int64)`, we should be able to trigger the same path as printing the dataframe can do.
```diff
diff --git a/polars/polars-core/src/datatypes/dtype.rs b/polars/polars-core/src/datatypes/dtype.rs
index d2b89aae08..9a41f8e26e 100644
--- a/polars/polars-core/src/datatypes/dtype.rs
+++ b/polars/polars-core/src/datatypes/dtype.rs
@@ -279,7 +279,7 @@ impl Display for DataType {
             }
             DataType::Duration(tu) => return write!(f, "duration[{tu}]"),
             DataType::Time => "time",
-            DataType::List(tp) => return write!(f, "list[{tp}]"),
+            DataType::List(tp) => return write!(f, "list[{tp}][test]"),
             #[cfg(feature = "object")]
             DataType::Object(s) => s,
             #[cfg(feature = "dtype-categorical")]
```


```python
>>> import polars as pl
>>> df = pl.DataFrame({"g": [[5, 6], [3, 4], [9, 8]]})
>>> df
shape: (3, 1)
┌─────────────────┐
│ g               │
│ ---             │
│ list[i64][test] │
╞═════════════════╡
│ [5, 6]          │
│ [3, 4]          │
│ [9, 8]          │
└─────────────────┘
```

**Comment by Mehrbod2002** [comment]
@ghuls 
I think it's not an relative compare between Dataframe which include Rust valid type to Datatype which include List , Object , Category ... and by this if you consider code as `string_repr` not initialize with Class so we lose inner Type on function and have to give it by ourself , and there is not choice .
If think just we have to replace some Rust type name on Datatype in `string_repr` such as i64 (Int64) , but as you can see we want to take Datatype method which Polars provide not Rust type which only dataframe can handle it . I think there are different .

**Comment by ghuls** [comment]
This will at least print the same dtype format string as the one printed by `print(df)`, ==> e.g `print(df.glimpse()0` ```diff
❯ git diff src/lib.rs
diff --git a/py-polars/src/lib.rs b/py-polars/src/lib.rs
index 1456092ca9..e636514e0c 100644
--- a/py-polars/src/lib.rs
+++ b/py-polars/src/lib.rs
@@ -108,7 +108,7 @@ fn dtype_cols(dtypes: Vec<Wrap<DataType>>) -> PyResult<dsl::PyExpr> {
 #[pyfunction]
 fn dtype_str_repr(dtype: Wrap<DataType>, inner: Option<Wrap<DataType>>) -> PyResult<String> {
     if let Some(inner) = inner {
-        return Ok(format!("List[{:?}]", inner.0));
+        return Ok(format!("list[{}]", inner.0.to_string()));
     }
     let dtype = dtype.0;
     Ok(dtype.to_string())
```

**Comment by Mehrbod2002** [comment]
> This will at least print the same dtype format string as the one printed by `print(df)`, ==> e.g `print(df.glimpse()0`
> 
> ```diff
> diff --git a/py-polars/src/lib.rs b/py-polars/src/lib.rs
> index 1456092ca9..e636514e0c 100644
> --- a/py-polars/src/lib.rs
> +++ b/py-polars/src/lib.rs
> @@ -108,7 +108,7 @@ fn dtype_cols(dtypes: Vec<Wrap<DataType>>) -> PyResult<dsl::PyExpr> {
>  #[pyfunction]
>  fn dtype_str_repr(dtype: Wrap<DataType>, inner: Option<Wrap<DataType>>) -> PyResult<String> {
>      if let Some(inner) = inner {
> -        return Ok(format!("List[{:?}]", inner.0));
> +        return Ok(format!("list[{}]", inner.0.to_string()));
>      }
>      let dtype = dtype.0;
>      Ok(dtype.to_string())
> ```

Will do it ASAP

**Comment by Mehrbod2002** [comment]
@ghuls Look at tests on [here](https://github.com/pola-rs/polars/actions/runs/3919448317/jobs/6700461859) which are provided on this PR .
trait `into` and `from` as I said provide Rust Type not Datatype .

**Comment by ghuls** [comment]
You need to fix the tests.
The only place `dtype_str_repr` is used is in `Dataframe.glimpse()` and it should print the dtypes in the same way as `print(df)` does.

**Comment by Mehrbod2002** [comment]
@ghuls Sorry I can see both of them is working as a same ! 
Can you mention which test ?

**Comment by Mehrbod2002** [comment]
@ghuls any new news ?

**Comment by Mehrbod2002** [comment]
@alexander-beedie Please consider to look this PR.
Thanks a lot .

File: py-polars/tests/unit/test_lists.py [hunk 1]
```diff
@@ -75,7 +75,7 @@ def test_dtype() -> None:
     a = pl.Series("a", [[1, 2, 3], [2, 5], [6, 7, 8, 9]])
     assert a.dtype == pl.List
     assert a.inner_dtype == pl.Int64
-    assert a.dtype.inner == pl.Int64  # type: ignore[attr-defined]
+    assert a.dtype.inner == pl.Int64  ### type: ignore[attr-defined]
```
Comment (stinodego) [inline py-polars/tests/unit/test_lists.py:78]: Quick drive-by comment: not sure why this change was introduced - please revert it. Same for line 269.

Comment (Mehrbod2002) [reply py-polars/tests/unit/test_lists.py:78]: It's failure in GitHub Action for it .
Here https://github.com/pola-rs/polars/commit/34ec4b4552b1baf5992920dc34dbf4bc109203dd

Comment (Mehrbod2002) [reply py-polars/tests/unit/test_lists.py:78]: Please consider look out these #6206
I did it that but mypy in GitHub Action stopped process , look out for commit please there

Comment (ritchie46) [reply py-polars/tests/unit/test_lists.py:78]: Then you have to fix that specific mypy lint. The current "commenting out" is not how this issue should be addressed.

What is the mypy lint you got?

Comment (Mehrbod2002) [reply py-polars/tests/unit/test_lists.py:78]: @ritchie46 
Sorry not familiar with mypy much ,
Just executed pre commit on reach this issue .
How can I config mypy for this issue ?

Comment (Mehrbod2002) [reply py-polars/tests/unit/test_lists.py]: @ritchie46 [here](https://github.com/pola-rs/polars/actions/runs/3937755230/jobs/6735643260) is error

**Comment by ghuls** [comment]
Locally I have it with: `return Ok(format!("list[{}]", inner.0));`

Does the following work fo you? It seems like the inner type is not stable.
```
>>> x = pl.List(pl.Utf8)
>>> x.string_repr()
'list[str]'
>>> y = pl.List(pl.UInt8)
>>> y.string_repr()
'list[u8]'
>>> x.string_repr()
'list[u8]'
```

**Comment by Mehrbod2002** [comment]
> Locally I have it with: `return Ok(format!("list[{}]", inner.0));`
> 
> Does the following work fo you? It seems like the inner type is not stable.
> 
> ```
> >>> x = pl.List(pl.Utf8)
> >>> x.string_repr()
> 'list[str]'
> >>> y = pl.List(pl.UInt8)
> >>> y.string_repr()
> 'list[u8]'
> >>> x.string_repr()
> 'list[u8]'
> ```

Because it's `@classmethod` if we are going to delete it will resolve ,
but non-list Type will become like this :
`pl.Utf8.string_repr(pl.Utf8)` works .

Can we create string_repr function new one for list ? 
For example string_repr_list ? 
Or is it acceptable to initial python class in `__new__` function ? don't think so but if it's possible could be great.

**Comment by ghuls** [comment]
pl.Datetime has the same problem:
```
>>> pl.Datetime("ns").string_repr()
'datetime[μs]'
```

**Comment by Mehrbod2002** [comment]
> pl.Datetime has the same problem:
> ```
> >>> pl.Datetime("ns").string_repr()
> 'datetime[μs]'
> ```

Please read my comment above ? 
Please read it it's solution 
Waiting for you command about it

**Comment by ghuls** [comment]
Superseeded by: https://github.com/pola-rs/polars/pull/6298

**Comment by Mehrbod2002** [comment]
> Superseeded by: https://github.com/pola-rs/polars/pull/6298

How about this PR ?

**Comment by Mehrbod2002** [comment]
@ghuls 
No issue with this PR then to print inner for list , yes ?

**Comment by ghuls** [comment]
@Mehrbod2002 It print the inner dtype correctly if string_repr is called on the instance instead of the class. Only some Polars dtypes are instantiated, so depending the dtype is an instance or the class we need to call `string_repr` slightly different. As we only need it for glimpse, this is fine for now.
https://github.com/pola-rs/polars/pull/6298/commits/56d2947b9a3a8f9a933a0173bb15720ed9c2dcfb#diff-0a0c19eef131f26f17dab6b0f66bb23abd81cb696e646626e2045794e274f6beL2733-R2738

**Comment by Mehrbod2002** [comment]
@ghuls 
Anyway if you want to do it assign yourself your reply to my commands to don't waste onther programmer time,  I supposed this community to solve and welcomed for new programmers or contributors.  Next time mention you want to do it yourself , I could do that and mention problem but you did as yourself without any response from last day
